lib/server.go

```diff
@@ -2058,25 +2127,52 @@ func (srv *Server) _handleGetTransactions(pp *Peer, msg *MsgDeSoGetTransactions)
 	pp.AddDeSoMessage(msg, true /*inbound*/)
 }
 
-func (srv *Server) ProcessSingleTxnWithChainLock(
-	pp *Peer, txn *MsgDeSoTxn) ([]*MempoolTx, error) {
+func (srv *Server) ProcessSingleTxnWithChainLock(pp *Peer, txn *MsgDeSoTxn) ([]*MsgDeSoTxn, error) {
 	// Lock the chain for reading so that transactions don't shift under our feet
 	// when processing this bundle. Not doing this could cause us to miss transactions
 	// erroneously.
 	//
 	// TODO(performance): We should probably do this less frequently.
 	srv.blockchain.ChainLock.RLock()
-	defer func() {
-		srv.blockchain.ChainLock.RUnlock()
-	}()
+	defer srv.blockchain.ChainLock.RUnlock()
+
 	// Note we set rateLimit=false because we have a global minimum txn fee that should
 	// prevent spam on its own.
-	return srv.mempool.ProcessTransaction(
-		txn, true /*allowUnconnectedTxn*/, false, /*rateLimit*/
-		pp.ID, true /*verifySignatures*/)
+
+	// Only attempt to add the transaction to the PoW mempool if we're on the
+	// PoW protocol. If we're on the PoW protocol, then we use the PoW mempool's,
+	// txn validity checks to signal whether the txn has been added or not. The PoW
+	// mempool has stricter txn validity checks than the PoW mempool, so this works
+	// out conveniently, as it allows us to always add a txn to the PoS mempool.
+	blockHeight := srv.blockchain.blockTip().Height
+	if blockHeight < srv.blockchain.params.ForkHeights.BalanceModelBlockHeight {
+		_, err := srv.mempool.ProcessTransaction(
+			txn,
+			true,  /*allowUnconnectedTxn*/
+			false, /*rateLimit*/
+			pp.ID,
+			true, /*verifySignatures*/
+		)
+
+		// If we're on the PoW chain, and the txn doesn't pass the PoW mempool's validity checks, then
+		// it's an invalid txn.
+		if err != nil {
+			return nil, errors.Wrapf(err, "Server.ProcessSingleTxnWithChainLock: Problem adding transaction to PoW mempool: ")
+		}
+	}
+
+	// Regardless of the consensus protocol we're running (PoW or PoS), we use the PoS mempool's to house all
+	// mempool txns. If a txn can't make it into the PoS mempool, which uses a looser unspent balance check for
+	// the the transactor, then it must be invalid.
+	if err := srv.posMempool.AddTransaction(NewMempoolTransaction(txn, uint64(time.Now().UnixMicro())), true); err != nil {
```

Comment:
Okay this isn't the right place for this comment, but wanted to get it out while it's in my head. Currently, we call `LoadTxnsFromDB` on the legacy mempool in `NewDeSoMempool`. When we start up a node for the first time with the PoSMempool, do we want to load these transation into the PoSMempool as well? I guess it's not a huge deal since those transactions should make it into a block relatively quickly, but we may want the two to be in lockstep.

Reply:
Good question. I didn't have a great solution for it. My thought is that it becomes a self-balancing thing once 2-3 blocks have mined. When a node first upgrades its code to the latest code and initializes the two mempools, the PoW mempool will hold all of the persisted txns while the PoS mempool will be empty. Then as txns are received organically, and as txns are mined into blocks, the two mempools will end up with the same contents. Both mempools will also dump their contents to disk so that they will maintain parity on future node restarts.

Reply:
yeah I like this better than trying to have some condition for the situation like:
if the PosMempool hasn't been dumped before, then load it with the transactions from the PowMempool. So I think we're okay to ignore this comment.